src/data/nat/multiplicity.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2019 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import data.nat.choose ring_theory.multiplicity data.nat.modeq algebra.gcd_domain
+/-!
+
+# Natural number multiplicity
+
+This file contains lemmas about the multiplicity function (the maximum prime power divding a number).
+
+# Main results
+
+There are natural number versions of some basic lemmas about multiplicity.
+
+There are also lemmas about the multiplicity of primes in factorials and in binomial coefficients.
+-/
+
+open finset nat multiplicity
+
+namespace nat
+
+/-- The multiplicity of a divisor `m` of `n`, is the cardinality of the set of
+  positive natural numbers `i` such that `p ^ i` divides `n`. The set is expressed
+  by filtering `Ico 1 b` where `b` is any bound at least `n` -/
+lemma multiplicity_eq_card_pow_dvd {m n b : ℕ} (hm1 : m ≠ 1) (hn0 : 0 < n) (hb : n ≤ b):
+  multiplicity m n = ↑((finset.Ico 1 b).filter (λ i, m ^ i ∣ n)).card :=
+calc multiplicity m n = ↑(Ico 1 $ ((multiplicity m n).get (finite_nat_iff.2 ⟨hm1, hn0⟩) + 1)).card :
+  by simp
+... = ↑((finset.Ico 1 b).filter (λ i, m ^ i ∣ n)).card : congr_arg coe $ congr_arg card $
+  finset.ext.2 $ λ i,
+  have hmn : ¬ m ^ n ∣ n,
+    from if hm0 : m = 0
+    then λ _, by cases n; simp [*, lt_irrefl, nat.pow_succ] at *
+    else mt (le_of_dvd hn0) (not_le_of_gt $ lt_pow_self
+      (lt_of_le_of_ne (nat.pos_of_ne_zero hm0) hm1.symm) _),
+  ⟨λ hi, begin
+      simp only [Ico.mem, mem_filter, lt_succ_iff] at *,
+      exact ⟨⟨hi.1, lt_of_le_of_lt hi.2 $
+        lt_of_lt_of_le (by rw [← enat.coe_lt_coe, enat.coe_get, multiplicity_lt_iff_neg_dvd,
+            nat.pow_eq_pow]; exact hmn)
+          hb⟩,
+        by rw [← nat.pow_eq_pow, pow_dvd_iff_le_multiplicity];
+          rw [← @enat.coe_le_coe i, enat.coe_get] at hi; exact hi.2⟩
+    end,
+  begin
+    simp only [Ico.mem, mem_filter, lt_succ_iff, and_imp, true_and] { contextual := tt },
+    assume h1i hib hmin,
+    rwa [← enat.coe_le_coe, enat.coe_get, ← pow_dvd_iff_le_multiplicity, nat.pow_eq_pow]
+  end⟩
+
+namespace prime
+
+lemma multiplicity_one {p : ℕ} (hp : p.prime) :
```

Comment:
Should these be simp-lemmas?

Reply:
They have a hypothesis so I don't think so.

Reply:
I think they're reasonable simp lemmas. They'll work if there's a proof of `p.prime` in the simp set. But it might be better to declare them as a simp set, and not add them to the default set.